File: kt/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/SyntaxFacts.kt
package dev.phytolizer.minsk.analysis.syntax

object SyntaxFacts {
    fun binaryOperatorPrecedence(kind: SyntaxKind): Int = when (kind) {
        SyntaxKind.StarToken, SyntaxKind.SlashToken -> 5
        SyntaxKind.PlusToken, SyntaxKind.MinusToken -> 4

        SyntaxKind.EqualsEqualsToken, SyntaxKind.BangEqualsToken,
        SyntaxKind.LessToken, SyntaxKind.LessOrEqualsToken,
        SyntaxKind.GreaterToken, SyntaxKind.GreaterOrEqualsToken,
        -> 3

        SyntaxKind.AmpersandAmpersandToken -> 2
        SyntaxKind.PipePipeToken -> 1
        else -> 0
    }

    fun unaryOperatorPrecedence(kind: SyntaxKind): Int = when (kind) {
        SyntaxKind.PlusToken, SyntaxKind.MinusToken, SyntaxKind.BangToken -> 6
        else -> 0
    }

    fun keywordKind(text: String): SyntaxKind = when (text) {
        "else" -> SyntaxKind.ElseKeyword
        "false" -> SyntaxKind.FalseKeyword
        "if" -> SyntaxKind.IfKeyword
        "let" -> SyntaxKind.LetKeyword
        "true" -> SyntaxKind.TrueKeyword
        "var" -> SyntaxKind.VarKeyword
        else -> SyntaxKind.IdentifierToken
    }

    fun binaryOperators(): List<SyntaxKind> = SyntaxKind.values().filter { binaryOperatorPrecedence(it) > 0 }
    fun unaryOperators(): List<SyntaxKind> = SyntaxKind.values().filter { unaryOperatorPrecedence(it) > 0 }

    fun getText(kind: SyntaxKind): String? = when (kind) {
        SyntaxKind.PlusToken -> "+"
        SyntaxKind.MinusToken -> "-"
        SyntaxKind.StarToken -> "*"
        SyntaxKind.SlashToken -> "/"
        SyntaxKind.OpenParenthesisToken -> "("
        SyntaxKind.CloseParenthesisToken -> ")"
        SyntaxKind.OpenBraceToken -> "{"
        SyntaxKind.CloseBraceToken -> "}"
        SyntaxKind.BangToken -> "!"
        SyntaxKind.AmpersandAmpersandToken -> "&&"
        SyntaxKind.PipePipeToken -> "||"
        SyntaxKind.BangEqualsToken -> "!="
        SyntaxKind.EqualsEqualsToken -> "=="
        SyntaxKind.EqualsToken -> "="
        SyntaxKind.LessToken -> "<"
        SyntaxKind.LessOrEqualsToken -> "<="
        SyntaxKind.GreaterToken -> ">"
        SyntaxKind.GreaterOrEqualsToken -> ">="
        SyntaxKind.ElseKeyword -> "else"
        SyntaxKind.FalseKeyword -> "false"
        SyntaxKind.IfKeyword -> "if"
        SyntaxKind.LetKeyword -> "let"
        SyntaxKind.TrueKeyword -> "true"
        SyntaxKind.VarKeyword -> "var"
        else -> null
    }
}

File: nim/src/minskpkg/codeAnalysis/syntax/syntaxFacts.nim
import std/options
import std/sequtils

import syntaxKind

proc binaryOperatorPrecedence*(kind: SyntaxKind): int =
  case kind
  of SyntaxKind.StarToken, SyntaxKind.SlashToken:
    5
  of SyntaxKind.PlusToken, SyntaxKind.MinusToken:
    4
  of SyntaxKind.EqualsEqualsToken, SyntaxKind.BangEqualsToken:
    3
  of SyntaxKind.AmpersandAmpersandToken:
    2
  of SyntaxKind.PipePipeToken:
    1
  else:
    0

proc unaryOperatorPrecedence*(kind: SyntaxKind): int =
  case kind
  of SyntaxKind.PlusToken,
      SyntaxKind.MinusToken,
      SyntaxKind.BangToken:
    6
  else:
    0

proc keywordKind*(text: string): SyntaxKind =
  case text
  of "true":
    SyntaxKind.TrueKeyword
  of "false":
    SyntaxKind.FalseKeyword
  else:
    SyntaxKind.IdentifierToken

proc text*(kind: SyntaxKind): Option[string] =
  case kind
  of SyntaxKind.PlusToken:
    "+".some
  of SyntaxKind.MinusToken:
    "-".some
  of SyntaxKind.StarToken:
    "*".some
  of SyntaxKind.SlashToken:
    "/".some
  of SyntaxKind.BangToken:
    "!".some
  of SyntaxKind.EqualsToken:
    "=".some
  of SyntaxKind.AmpersandAmpersandToken:
    "&&".some
  of SyntaxKind.PipePipeToken:
    "||".some
  of SyntaxKind.EqualsEqualsToken:
    "==".some
  of SyntaxKind.BangEqualsToken:
    "!=".some
  of SyntaxKind.OpenParenthesisToken:
    "(".some
  of SyntaxKind.CloseParenthesisToken:
    ")".some
  of SyntaxKind.FalseKeyword:
    "false".some
  of SyntaxKind.TrueKeyword:
    "true".some
  else:
    none(string)

proc binaryOperators*(): seq[SyntaxKind] =
  SyntaxKind.toSeq.filter(
    proc(kind: SyntaxKind): bool =
      kind.binaryOperatorPrecedence > 0
  )

proc unaryOperators*(): seq[SyntaxKind] =
  SyntaxKind.toSeq.filter(
    proc(kind: SyntaxKind): bool =
      kind.unaryOperatorPrecedence > 0
  )
